### PR TITLE
Ports TGs way of handling bullet_act on shooting targets. Should fix the armory crash

### LINKED ISSUE #11037: [META] Crashing near science
**Problem Description**:
Me and others have crashed multiple times now when approaching science on metastation

**What did you expect to happen**: 
Not to crash

**What happened instead**:
CRASH

**Why is this bad/What are the consequences:**
It crashes you and takes you out of the game for 30+ seconds (yes ads that don't load)

**Steps to reproduce the problem**:
Walk near science. It happend to me in the hallway to escape

### LINKED ISSUE #11142: Crash Tiles at the Armory
**Problem Description**:
The crashing hallway tiles on metastation seem to now happen on the Cyberiad, though in the brig armory area. The small spot with the vendor and the secure armory itself. Once again random where exactly it happens. Happened on two separate rounds, does not always happen where it did previously or even at all.

**What did you expect to happen**: 
I expected to not crash when entering the armory.

**What happened instead**:
Myself and 3-4 other officers crashed when entering the armory.

**Why is this bad/What are the consequences:**
Number of reasons. Being chased by an antag, trying to gear up for a threat, generally existing to go recharge your gear.

**Steps to reproduce the problem**:
1. Be sec.
2. Go into the armory room.
3. Walk around on the tiles until you possibly crash.

**When did the problem start happening**:
About a few days ago is when I noticed it when I started playing security again.

**Possibly related stuff (which gamemode was it? What were you doing at the time? Was anything else out of the ordinary happening?)**:
Nothing relevant.

### PR DESCRIPTION
**Pull request by farie82**
## What Does This PR Do
Ports the way TG handles targets. I've changed some things up that didn't make much sense. I'll list them in the comments.
Also did a small refactor

Fixes #11142
Fixes #11037
Alternative to #12928

## Why It's Good For The Game
Won't crash clients now I hope

## Changelog
:cl:
fix: Fixes the armory crash
/:cl:

### CODE REVIEW

File: code/game/objects/items/shooting_range.dm [hunk 1]
```diff
@@ -77,101 +76,36 @@
...
+#define DECALTYPE_BULLET 2
+
+/obj/item/target/bullet_act(obj/item/projectile/P)
+	var/p_x = P.p_x + pick(0,0,0,0,0,-1,1) // really ugly way of coding "sometimes offset P.p_x!"
+	var/p_y = P.p_y + pick(0,0,0,0,0,-1,1)
+	var/decaltype = DECALTYPE_SCORCH
+	if(istype(P, /obj/item/projectile/bullet))
+		decaltype = DECALTYPE_BULLET
+	var/icon/C = icon(icon, icon_state)
+	if(C.GetPixel(p_x, p_y) && LAZYLEN(overlays) <= 35) // if the located pixel isn't blank (null)
```
Comment (farie82) [inline code/game/objects/items/shooting_range.dm:89]: `if(C.GetPixel(p_x, p_y) && P.original == src && overlays.len <= 35)` is the check TG has. The original check didn't make much sense to me so I kept it out.

File: code/game/objects/items/shooting_range.dm [hunk 2]
```diff
@@ -77,101 +76,36 @@
...
-	virtualIcon = new(icon, icon_state)
-
-	if( virtualIcon.GetPixel(p_x, p_y) ) // if the located pixel isn't blank (null)
-
-		hp -= Proj.damage
+#define DECALTYPE_SCORCH 1
+#define DECALTYPE_BULLET 2
+
+/obj/item/target/bullet_act(obj/item/projectile/P)
+	var/p_x = P.p_x + pick(0,0,0,0,0,-1,1) // really ugly way of coding "sometimes offset P.p_x!"
```
Comment (farie82) [inline code/game/objects/items/shooting_range.dm:83]: They check for a riot dart or such at the start here. I've left that out for now

File: code/game/objects/items/shooting_range.dm [hunk 3]
```diff
@@ -77,101 +76,36 @@
...
-			// offset correction
-			bmark.pixel_x--
-			bmark.pixel_y--
-
-			if(Proj.damage >= 20 || istype(Proj, /obj/item/projectile/beam/practice))
-				bmark.icon_state = "scorch"
-				bmark.dir = pick(NORTH,SOUTH,EAST,WEST) // random scorch design
-
-
+			return
```
Comment (farie82) [inline code/game/objects/items/shooting_range.dm:94]: There was no return on the TG version. qdel will still finish this proc so yeah an early return is best

File: code/game/objects/items/shooting_range.dm [hunk 4]
```diff
@@ -77,101 +76,36 @@
...
+		var/image/bullet_hole = image('icons/effects/effects.dmi', "scorch", OBJ_LAYER + 0.5)
+		bullet_hole.pixel_x = p_x - 1 //offset correction
+		bullet_hole.pixel_y = p_y - 1
+		if(decaltype == DECALTYPE_SCORCH)
+			if(P.damage >= 20 || istype(P, /obj/item/projectile/beam/practice))
+				bullet_hole.setDir(pick(NORTH,SOUTH,EAST,WEST))// random scorch design. light_scorch does not have different directions
```
Comment (farie82) [inline code/game/objects/items/shooting_range.dm:100]: This one was duped in TG before line 99. Issue is that only scorch has 4 directions and light_scorch does not. TG has the same icon.

**Comment by KasparoVy** [comment]
Badass.

**Comment by BakKeener** [comment]
finally no more armory crashes!

**Pull request merged**